katsdpdata/prod_handler.py

```diff
@@ -1,53 +1,619 @@
+        """True when files have been staged for transfer"""
+        return bool(self._staged_for_transfer)
+
+    def upload_size(self):
+        """This product's contribution to the upload size"""
+        return sum(
+            os.path.getsize(f) for f in self._staged_for_transfer
+            if os.path.isfile(f))
+
+    def update_state(self, transition):
+        """The product state machine based on state transitions
+        TODO: There are still some state changes baked into the metadata
+        TODO: handlers, but which are for explicivity duplicated here.
+        """
+        mh = self.mh()
+        current_state = mh.get_state()
+        if transition == 'TRANSFER_DONE':
+            if current_state == 'TRANSFERRING':
+                mh.set_product_status('RECEIVED')
+            elif current_state == 'RESTAGING':
+                mh.set_product_status('RESTAGED')
+            self.metadata_transfer_complete()
+        elif transition == 'PRODUCT_DETECTED':
+            if not current_state:
+                self.metadata_when_created()
+                mh.set_product_status('CREATED')
+            elif current_state == 'ARCHIVED':
+                self.metadata_when_created()
+                mh.set_product_status('RECREATED')
+        elif transition == 'TRANSFER_STARTED':
+            if not current_state:
+                self.metadata_when_created()
+                mh.set_product_status('TRANSFERRING')
+            elif current_state == 'CREATED':
+                mh.set_product_status('TRANSFERRING')
+            elif current_state in ['RECREATED', 'ARCHIVED']:
+                mh.set_product_status('RESTAGING')
+        elif transition == 'FAILED':
+            mh.set_product_status('FAILED')
+
+    def metadata_transfer_complete(self):
+        mh = self.mh()
+        met = mh.get_prod_met()
+        met_bucket = self.get_bucket_stats()
+        mh.add_bucket_stats(met, met_bucket)
+
+    def metadata_when_created(self):
+        mh = self.mh()
+        mh.create_core_met()
+
+    def bucket_name(self):
+        """Get the bucket name from the product_path"""
+        if not self.product_path:
+            return None
+        return [x for x in self.product_path.split(os.sep) if x][-1]
+
+    def get_bucket_stats(self):
+        """Get the bucket stats from S3"""
+        return {}
+        # TODO: Update the kat_core then these fields can be added
+        bucket_name = self.bucket_name()
+        try:
+            s3_conn = get_s3_connection(self.boto_dict)
+            bucket = s3_conn.get_bucket(bucket_name)
+        except boto.exception.S3ResponseError as e:
+            logger.error(
+                f'Could not get bucket stats for {bucket_name}. '
+                f'It does not seem to exist.')
+            return None
+        met_bucket = {
+            'prefix': bucket.name,
+            'BucketOwner': bucket.get_acl().owner.display_name,
+            'size': sum([k.size for k in bucket.get_all_keys()]),
```

Comment:
I would test this on a real observation to see how long it takes.

This observation 1625673674 is about 5TB.

Reply:
Ok, so I figured out the get_all_keys() just gets the first 1000 keys and to get an iterator over the entire bucket's keys the list method needs to be called. Which I timed:
```
>>> start = time.time()
>>> total_size = sum([k.size for k in bucket.list()])
>>> stop = time.time()
>>> print(f'It took {stop - start} to calculate the total of {total_size}')
It took 57.252509355545044 to calculate the total of 6248862155264
>>> 
```
Looks like the same ballpark on the second run: 
`It took 52.16948318481445 to calculate the total of 6248862155264.` 
I'll have to update this, but @ctgschollar , do you think a maximum delay of 1 minute is an acceptable delay for an observation until we move over to the ceph_core approach?